wandb/beta/workflows.py

```diff
@@ -0,0 +1,192 @@
+    elif isinstance(path_or_obj, str):
+        if os.path.isdir(path_or_obj):
+            return artifact.add_dir(path_or_obj)
+        elif os.path.isfile(path_or_obj):
+            return artifact.add_file(path_or_obj)
+        else:
+            import json
+
+            with artifact.new_file(name) as f:
+                f.write(json.dumps(path_or_obj, sort_keys=True))
+    else:
+        raise ValueError(
+            f"Expected `path_or_obj` to be instance of `ArtifactEntry`, `WBValue`, or `str, found {type(path_or_obj)}"
+        )
+
+
+def _log_artifact_version(
+    name: str,
+    type: str,
+    entries: Dict[str, Union[str, ArtifactEntry, data_types.WBValue]],
+    aliases: Optional[Union[str, List[str]]] = None,
+    description: Optional[str] = None,
+    metadata: Optional[dict] = None,
+    project: Optional[str] = None,
+    scope_project: Optional[bool] = None,
+    job_type: str = "auto",
+) -> ArtifactInterface:
+    if wandb.run is None:
+        run = wandb.init(
+            project=project, job_type=job_type, settings=wandb.Settings(silent="true")
+        )
+    else:
+        run = wandb.run
+
+    if not scope_project:
+        name = f"{name}-{run.id}"
+
+    if metadata is None:
+        metadata = {}
+
+    # hack for demo purposes
+    # if getattr(run, "history", None):
+    #     metadata.update({"__wb_log_step__": wandb.run.history._step})
+
+    art = wandb.Artifact(name, type, description, metadata, False, None)
+
+    for path in entries:
+        _add_any(art, entries[path], path)
+
+    # "latest" should always be present as an alias in every new logged artifact version.
+    if aliases is None:
+        aliases = []
+    elif isinstance(aliases, str):
+        aliases = [aliases]
+
+    if "latest" not in aliases:
+        aliases.append("latest")
+
+    run.log_artifact(art, aliases=aliases)
+
+    return art
+
+
+def log_model(
+    model_obj: Any,
+    name: str = "model",
+    aliases: Optional[Union[str, List[str]]] = None,
+    description: Optional[str] = None,
+    metadata: Optional[dict] = None,
+    project: Optional[str] = None,
+    scope_project: Optional[bool] = None,
+) -> "SavedModel":
+    model = data_types.SavedModel.init(model_obj)
+    _ = _log_artifact_version(
+        name=name,
+        type="model",
+        entries={"index": model,},
+        aliases=aliases,
+        description=description,
+        metadata=metadata,
+        project=project,
+        scope_project=scope_project,
+        job_type="log_model",
+    )
+
+    # TODO: handle offline mode appropriately.
+    return model
+
+
+def use_model(alias: str) -> "SavedModel":
+    parts = alias.split(":")
+    if len(parts) == 1:
+        alias += "latest"
+
+    # Returns a SavedModel instance
+    if wandb.run:
+        run = wandb.run
+        artifact = run.use_artifact(alias)
+        sm = artifact.get("index")
+        return sm
+    else:
+        raise ValueError(
```

Comment:
Yeah, this is interesting -- let's think about this. The value of `use_model` is two-fold:
1. For us, it binds the model artifact as an input to a run, which would ultimately allow us to do evaluation workflows in the UI. 
2. For the user, they get deserialization out of the box in an easy and framework-agnostic way, so they can do training and evaluation. 

Point 2 is always accomplished whether we start a run or not, so let's focus on 1. 

If we start a run for the user, 1 is accomplished _as long as the user does not do_ `wandb.init()` after the `use_model` call(s). If they do, however, the model artifacts won't be bound to the run we're interested in. 

If we throw an error (and make the user always init a run before `use_model`), I think this would ensure that 1 is accomplished correctly -- that is, the model artifacts are almost always bound to the same run that logs model metrics (i.e for the evaluation use-case).